pkg/capacity/manager.go

```diff
@@ -6,68 +6,60 @@ import (
 	"github.com/codeready-toolchain/host-operator/pkg/counter"
 	"github.com/codeready-toolchain/toolchain-common/pkg/cluster"
 	"github.com/codeready-toolchain/toolchain-common/pkg/condition"
+	"sigs.k8s.io/controller-runtime/pkg/log"
 
-	"k8s.io/apimachinery/pkg/types"
 	runtimeclient "sigs.k8s.io/controller-runtime/pkg/client"
 )
 
 type (
-	spaceProvisionerConfigPredicate func(*toolchainv1alpha1.SpaceProvisionerConfig) bool
-)
+	spaceProvisionerConfigPredicate func(context.Context, *toolchainv1alpha1.SpaceProvisionerConfig) bool
 
-func hasNotReachedMaxNumberOfSpacesThreshold(counts counter.Counts) spaceProvisionerConfigPredicate {
-	return func(spc *toolchainv1alpha1.SpaceProvisionerConfig) bool {
-		numberOfSpaces := uint(counts.SpacesPerClusterCounts[spc.Spec.ToolchainCluster])
-		threshold := spc.Spec.CapacityThresholds.MaxNumberOfSpaces
-		return threshold == 0 || numberOfSpaces < threshold
-	}
-}
+	// SpaceCountGetter is a function useful for mocking the counts cache and can be passed
+	// to the NewClusterManager function. The returned tuple represents the actual number
+	// of the spaces found for given cluster and the fact whether the value was actually found or not
+	// (similar to the return value of the map lookup).
+	SpaceCountGetter func(ctx context.Context, clusterName string) (int, bool)
+)
 
-func hasEnoughMemoryCapacity(status *toolchainv1alpha1.ToolchainStatus) spaceProvisionerConfigPredicate {
-	return func(spc *toolchainv1alpha1.SpaceProvisionerConfig) bool {
-		threshold := spc.Spec.CapacityThresholds.MaxMemoryUtilizationPercent
-		if threshold == 0 {
-			return true
+// Even though the readiness status of the SpaceProvisionerConfig is based (in part) on the space count
+// we do another check here using the fresh data from the counts cache. This is to ensure we don't
+// overcommit spaces to the clusters.
+//
+// NOTE: THE RETURNED FUNCTION ACTUALLY UPDATES THE SPC IT RECEIVES WITH THE LATEST SPACE COUNT FROM THE
+// CACHE. THIS CAN BE TAKEN ADVANTAGE OF LATER WHEN FURTHER PROCESSING THE CANDIDATE SPCS.
+func recheckHasNotReachedSpaceCountThreshold(getSpaceCount SpaceCountGetter) spaceProvisionerConfigPredicate {
+	return func(ctx context.Context, spc *toolchainv1alpha1.SpaceProvisionerConfig) bool {
+		var spaceCount int
+		if spc.Status.ConsumedCapacity != nil {
+			spaceCount = spc.Status.ConsumedCapacity.SpaceCount
 		}
-		for _, memberStatus := range status.Status.Members {
-			if memberStatus.ClusterName == spc.Spec.ToolchainCluster {
-				return hasMemberEnoughMemoryCapacity(memberStatus, threshold)
+		cachedCount, ok := getSpaceCount(ctx, spc.Spec.ToolchainCluster)
+		if ok {
+			// the cachedCount is always going to be fresher than (or as fresh as) what's in the SPC
+			spaceCount = cachedCount
+			if spc.Status.ConsumedCapacity == nil {
+				spc.Status.ConsumedCapacity = &toolchainv1alpha1.ConsumedCapacity{}
 			}
+			spc.Status.ConsumedCapacity.SpaceCount = spaceCount
```

Comment:
updating status from different controllers is an anti-pattern. Only the controller of the primary resource should update the status of the resource

Reply:
still, let's not change the retrieved object if we don't want to update the actual resource. If we want to keep the values somewhere, then store it in a different type/variable.
Out of curiosity, what it is used for?

Reply:
I redid this a bit, moving the update to a more logical place where I think it will not raise eyebrows. Please take another look.